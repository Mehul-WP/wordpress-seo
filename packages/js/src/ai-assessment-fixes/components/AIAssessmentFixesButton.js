import PropTypes from "prop-types";
import { __ } from "@wordpress/i18n";
import { useCallback, useRef } from "@wordpress/element";
import { doAction } from "@wordpress/hooks";
import { useSelect, useDispatch } from "@wordpress/data";

/* Yoast dependencies */
import { colors } from "@yoast/style-guide";
import { IconAIFixesButton } from "@yoast/components";
import { Modal, useToggleState } from "@yoast/ui-library";
import { Paper } from "yoastseo";

/* Internal dependencies */
import { ModalContent } from "../../ai-generator/components/modal-content";

const SparklesIcon = ( { pressed } ) => {
	return (
		<>
			<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
				<path
					d="M3.33284 2V4.66667M1.99951 3.33333H4.66618M3.99951 11.3333V14M2.66618 12.6667H5.33284M8.66618 2L10.19 6.57143L13.9995 8L10.19 9.42857L8.66618 14L7.14237 9.42857L3.33284 8L7.14237 6.57143L8.66618 2Z"
					stroke={ pressed ? "white" : "url(#paint0_linear_1208_188)" } strokeWidth="1.33333" strokeLinecap="round"
					strokeLinejoin="round"
				/>
				<defs>
					<linearGradient
						id="paint0_linear_1208_188" x1="0" y1="0" x2="16" y2="16"
						gradientUnits="userSpaceOnUse"
					>
						<stop stopColor="#A61E69" />
						<stop offset="1" stopColor="#3B82F6" />
					</linearGradient>
				</defs>
			</svg>
		</>
	);
};

/**
 * The AI Assessment Fixes button component.
 *
 * @param {string} id The assessment ID for which the AI fixes should be applied to.
 * @param {boolean} isPremium Whether the premium add-on is active.
 *
 * @returns {JSX.Element} The AI Assessment Fixes button.
 */
const AIAssessmentFixesButton = ( { id, isPremium } ) => {
	const aiFixesId = id + "AIFixes";
	const ariaLabel = __( "Fix with AI", "wordpress-seo" );
	const [ isModalOpen, , , setIsModalOpenTrue, setIsModalOpenFalse ] = useToggleState( false );
	const activeAIButtonId = useSelect( select => select( "yoast-seo/editor" ).getActiveAIFixesButton(), [] );
	const activeMarker = useSelect( select => select( "yoast-seo/editor" ).getActiveMarker(), [] );
	const { setActiveAIFixesButton, setActiveMarker, setMarkerPauseStatus } = useDispatch( "yoast-seo/editor" );
	const focusElementRef = useRef( null );

	/**
	 * Handles the button press state.
	 * @returns {void}
	 */
	const handlePressedButton = () => {
		// If there is an active marker when the AI fixes button is clicked, remove it.
		if ( activeMarker ) {
			setActiveMarker( null );
			setMarkerPauseStatus( false );
			// Remove highlighting from the editor.
			window.YoastSEO.analysis.applyMarks( new Paper( "", {} ), [] );
		}

		/* If the current pressed button ID is the same as the active AI button id,
		we want to set the active AI button to null. Otherwise, update the active AI button ID. */
		if ( aiFixesId === activeAIButtonId ) {
			setActiveAIFixesButton( null );
		} else {
			setActiveAIFixesButton( aiFixesId );
		}
	};

	const handleClick = useCallback( () => {
		if ( isPremium ) {
			doAction( "yoast.ai.fixAssessments", aiFixesId );
			/* Only handle the pressed button state in Premium.
			We don't want to change the background color of the button and other styling when it's pressed in Free.
			This is because clicking on the button in Free will open the modal, and the button will not be in a pressed state. */
			handlePressedButton();
		} else {
			setIsModalOpenTrue();
		}
	}, [ handlePressedButton, setIsModalOpenTrue ] );

	// The button is pressed when the active AI button id is the same as the current button id.
	const isButtonPressed = activeAIButtonId === aiFixesId;

	// Don't show the tooltip when the button is pressed.
	const className = isButtonPressed ? "" : "yoast-tooltip yoast-tooltip-w";

	return (
		<>
			<IconAIFixesButton
				onClick={ handleClick }
				ariaLabel={ ariaLabel }
				id={ aiFixesId }
				className={ className }
				pressed={ isButtonPressed }
			>
				<SparklesIcon pressed={ isButtonPressed } />
				{
					// We put the logic for the Upsell component in place.
					// The Modal below is only a placeholder/mock. When we have the design for the real upsell, the modal should be replaced.
					isModalOpen && <Modal className="yst-introduction-modal" isOpen={ isModalOpen } onClose={ setIsModalOpenFalse } initialFocus={ focusElementRef }>
						<Modal.Panel className="yst-max-w-lg yst-p-0 yst-rounded-3xl yst-introduction-modal-panel">
							<ModalContent onClose={ setIsModalOpenFalse } focusElementRef={ focusElementRef } />
						</Modal.Panel>
					</Modal>
				}
			</IconAIFixesButton>
		</>
	);
};

AIAssessmentFixesButton.propTypes = {
	id: PropTypes.string.isRequired,
	isPremium: PropTypes.bool,
};

AIAssessmentFixesButton.defaultProps = {
	isPremium: false,
};

export default AIAssessmentFixesButton;

