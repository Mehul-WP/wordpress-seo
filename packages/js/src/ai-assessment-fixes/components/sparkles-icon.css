/* sparkles-icon.css */
svg path.stroke-gradient {
	stroke: url(#paint0_linear_1208_188);
}

svg path.stroke-white {
	stroke: white;
}
