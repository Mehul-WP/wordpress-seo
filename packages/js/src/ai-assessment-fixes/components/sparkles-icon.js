import propTypes from "prop-types";

/**
 * The AI Assessment Fixes button icon.
 * @param {boolean} pressed Whether the button is pressed.
 * @returns {JSX.Element} The AI Assessment Fixes button icon.
 */
export const SparklesIcon = ( { pressed = false } ) => {
	return (
		<>
			<svg width="16" height="17" viewBox="0 0 16 17" fill="none" xmlns="http://www.w3.org/2000/svg">
				<path
					d="M3.33284 2.96991V5.63658M1.99951 4.30324H4.66618M3.99951 12.3032V14.9699M2.66618 13.6366H5.33284M8.66618 2.96991L10.19 7.54134L13.9995 8.96991L10.19 10.3985L8.66618 14.9699L7.14237 10.3985L3.33284 8.96991L7.14237 7.54134L8.66618 2.96991Z"
					stroke={ pressed ? "white" : "url(#paint0_linear_1538_754)" } strokeWidth="1.33333" strokeLinecap="round"
					strokeLinejoin="round"/>
				<defs>
					<linearGradient id="paint0_linear_1538_754" x1="1.99951" y1="2.96991" x2="15.3308" y2="4.69764"
									gradientUnits="userSpaceOnUse">
						<stop stopColor="#A61E69"/>
						<stop offset="1" stopColor="#6366F1"/>
					</linearGradient>
				</defs>
			</svg>
		</>
	);
};

SparklesIcon.propTypes = {
	pressed: propTypes.bool,
};
