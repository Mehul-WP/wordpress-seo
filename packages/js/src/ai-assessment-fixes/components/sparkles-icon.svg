<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
        d="M3.33284 2V4.66667M1.99951 3.33333H4.66618M3.99951 11.3333V14M2.66618 12.6667H5.33284M8.66618 2L10.19 6.57143L13.9995 8L10.19 9.42857L8.66618 14L7.14237 9.42857L3.33284 8L7.14237 6.57143L8.66618 2Z"
        stroke="url(#paint0_linear_1208_188)"
        strokeWidth="1.33333"
        strokeLinecap="round"
        strokeLinejoin="round"
    />
    <defs>
        <linearGradient
            id="paint0_linear_1208_188"
            x1="0"
            y1="0"
            x2="16"
            y2="16"
            gradientUnits="userSpaceOnUse"
        >
            <stop stopColor="#A61E69" />
            <stop offset="1" stopColor="#3B82F6" />
        </linearGradient>
    </defs>
</svg>
