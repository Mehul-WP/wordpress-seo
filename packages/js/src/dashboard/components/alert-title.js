import { useContext } from "@wordpress/element";
import PropTypes from "prop-types";
import classNames from "classnames";
import { ExclamationCircleIcon } from "@heroicons/react/outline";
import { Title } from "@yoast/ui-library";
import { AlertsContext } from "../routes/alert-center";

/**
 *
 * @param {string} title The title of alerts.
 * @param {number} counts The count of the alerts.
 *
 * @returns {JSX.Element} The alert title element.
 */
export const AlertTitle = ( {
	title,
	counts = 0,
} ) => {
	const { Icon = ExclamationCircleIcon, iconClass = "" } = useContext( AlertsContext );

	return (
		<div className="yst-flex yst-justify-between yst-gap-2 yst-items-center">
			<Icon className={ classNames( "yst-w-6 yst-h-6", iconClass ) } />
			<Title className="yst-grow" as="h3" size="2">{ title } { counts > 0 && `(${counts})` }</Title>
		</div>
	);
};

AlertTitle.propTypes = {
	title: PropTypes.string,
	counts: PropTypes.number,
};
