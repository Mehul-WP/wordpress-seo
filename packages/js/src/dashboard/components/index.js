export { default as RouteLayout } from "./route-layout";
export { default as SidebarNavigation } from "./sidebar-navigation";
