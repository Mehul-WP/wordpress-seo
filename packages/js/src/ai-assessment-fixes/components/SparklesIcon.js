import React from 'react';
import { ReactComponent as SparklesIcon } from './sparkles-icon.svg';

const IconWithConditionalStyle = ({ pressed }) => {
    return (
        <div>
            <SparklesIcon pressed={pressed} />
        </div>
    );
};

export default IconWithConditionalStyle;
