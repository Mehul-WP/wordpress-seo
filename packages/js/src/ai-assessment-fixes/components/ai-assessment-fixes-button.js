import PropTypes from "prop-types";
import { __ } from "@wordpress/i18n";
import { useCallback, useRef, useState } from "@wordpress/element";
import { doAction } from "@wordpress/hooks";
import { useSelect, useDispatch } from "@wordpress/data";

/* Yoast dependencies */
import { IconAIFixesButton } from "@yoast/components";
import { Modal, useToggleState } from "@yoast/ui-library";
import { Paper } from "yoastseo";

/* Internal dependencies */
import { ModalContent } from "./modal-content";
import { SparklesIcon } from "./sparkles-icon";

/**
 * The AI Assessment Fixes button component.
 *
 * @param {string} id The assessment ID for which the AI fixes should be applied to.
 * @param {boolean} isPremium Whether the premium add-on is active.
 *
 * @returns {JSX.Element} The AI Assessment Fixes button.
 */
const AIAssessmentFixesButton = ( { id, isPremium } ) => {
	const aiFixesId = id + "AIFixes";
	const ariaLabel = __( "Optimize with AI", "wordpress-seo" );
	const [ isModalOpen, , , setIsModalOpenTrue, setIsModalOpenFalse ] = useToggleState( false );
	const activeAIButtonId = useSelect( select => select( "yoast-seo/editor" ).getActiveAIFixesButton(), [] );
	const activeMarker = useSelect( select => select( "yoast-seo/editor" ).getActiveMarker(), [] );
	const { setActiveAIFixesButton, setActiveMarker, setMarkerPauseStatus } = useDispatch( "yoast-seo/editor" );
	const focusElementRef = useRef( null );
	const [ buttonClass, setButtonClass ] = useState( "" );

	/**
	 * Handles the button press state.
	 * @returns {void}
	 */
	const handlePressedButton = () => {
		// If there is an active marker when the AI fixes button is clicked, remove it.
		if ( activeMarker ) {
			setActiveMarker( null );
			setMarkerPauseStatus( false );
			// Remove highlighting from the editor.
			window.YoastSEO.analysis.applyMarks( new Paper( "", {} ), [] );
		}

		/* If the current pressed button ID is the same as the active AI button id,
		we want to set the active AI button to null. Otherwise, update the active AI button ID. */
		if ( aiFixesId === activeAIButtonId ) {
			setActiveAIFixesButton( null );
		} else {
			setActiveAIFixesButton( aiFixesId );
		}

		// Dismiss the tooltip when the button is pressed.
		setButtonClass( "" );
	};

	const handleClick = useCallback( () => {
		if ( isPremium ) {
			doAction( "yoast.ai.fixAssessments", aiFixesId );
			/* Only handle the pressed button state in Premium.
			We don't want to change the background color of the button and other styling when it's pressed in Free.
			This is because clicking on the button in Free will open the modal, and the button will not be in a pressed state. */
			handlePressedButton();
		} else {
			setIsModalOpenTrue();
		}
	}, [ handlePressedButton, setIsModalOpenTrue ] );

	// The button is pressed when the active AI button id is the same as the current button id.
	const isButtonPressed = activeAIButtonId === aiFixesId;

	// Add tooltip classes on mouse enter and remove them on mouse leave.
	const handleMouseEnter = useCallback( () => {
		// Add tooltip classes on mouse enter
		setButtonClass( "yoast-tooltip yoast-tooltip-w" );
	}, [] );

	const handleMouseLeave = useCallback( () => {
		// Remove tooltip classes on mouse leave
		setButtonClass( "" );
	}, [] );

	return (
		<>
			<IconAIFixesButton
				onClick={ handleClick }
				ariaLabel={ ariaLabel }
				onMouseEnter={ handleMouseEnter }
				onMouseLeave={ handleMouseLeave }
				id={ aiFixesId }
				className={ `ai-button ${buttonClass}` }
				pressed={ isButtonPressed }
			>
				<SparklesIcon pressed={ isButtonPressed }/>
				{
					// We put the logic for the Upsell component in place.
					// The Modal below is only a placeholder/mock. When we have the design for the real upsell, the modal should be replaced.
					isModalOpen && <Modal className="yst-introduction-modal" isOpen={ isModalOpen } onClose={ setIsModalOpenFalse } initialFocus={ focusElementRef }>
						<Modal.Panel className="yst-max-w-lg yst-p-0 yst-rounded-3xl yst-introduction-modal-panel">
							<ModalContent onClose={ setIsModalOpenFalse } focusElementRef={ focusElementRef } />
						</Modal.Panel>
					</Modal>
				}
			</IconAIFixesButton>
		</>
	);
};

AIAssessmentFixesButton.propTypes = {
	id: PropTypes.string.isRequired,
	isPremium: PropTypes.bool,
};

AIAssessmentFixesButton.defaultProps = {
	isPremium: false,
};

export default AIAssessmentFixesButton;

