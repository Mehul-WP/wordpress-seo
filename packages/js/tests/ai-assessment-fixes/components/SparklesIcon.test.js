import { render } from "../../test-utils";
import { SparklesIcon } from "../../../src/ai-assessment-fixes/components/sparkles-icon";

const testCases = [
	{
		name: "pressed is false",
		pressed: false,
	},
	{
		name: "pressed is true",
		pressed: true,
	},
];

describe.each( testCases )( "SparklesIcon", ( { name, pressed } ) => {
	test( `should render the SparklesIcon component when ${name}`, () => {

		// Mock Math.random to return a constant value for the gradient id.
		jest.spyOn(global.Math, 'random').mockReturnValue("000kdmaht");

		const { container } = render( <SparklesIcon pressed={ pressed } /> );

		expect( container ).toMatchSnapshot();

		// Clean up the mock to ensure tests are completely isolated.
		global.Math.random.mockRestore();
	} );
} );
