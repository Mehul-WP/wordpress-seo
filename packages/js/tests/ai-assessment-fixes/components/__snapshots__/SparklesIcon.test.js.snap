// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SparklesIcon should render the SparklesIcon component when pressed is false 1`] = `
<div>
  <svg
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3.33284 2V4.66667M1.99951 3.33333H4.66618M3.99951 11.3333V14M2.66618 12.6667H5.33284M8.66618 2L10.19 6.57143L13.9995 8L10.19 9.42857L8.66618 14L7.14237 9.42857L3.33284 8L7.14237 6.57143L8.66618 2Z"
      stroke="url(#paint0_linear_1208_188)"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.33333"
    />
    <defs>
      <lineargradient
        gradientUnits="userSpaceOnUse"
        id="paint0_linear_1208_188"
        x1="0"
        x2="16"
        y1="0"
        y2="16"
      >
        <stop
          stop-color="#A61E69"
        />
        <stop
          offset="1"
          stop-color="#6366F1"
        />
      </lineargradient>
    </defs>
  </svg>
</div>
`;

exports[`SparklesIcon should render the SparklesIcon component when pressed is true 1`] = `
<div>
  <svg
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3.33284 2V4.66667M1.99951 3.33333H4.66618M3.99951 11.3333V14M2.66618 12.6667H5.33284M8.66618 2L10.19 6.57143L13.9995 8L10.19 9.42857L8.66618 14L7.14237 9.42857L3.33284 8L7.14237 6.57143L8.66618 2Z"
      stroke="white"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.33333"
    />
    <defs>
      <lineargradient
        gradientUnits="userSpaceOnUse"
        id="paint0_linear_1208_188"
        x1="0"
        x2="16"
        y1="0"
        y2="16"
      >
        <stop
          stop-color="#A61E69"
        />
        <stop
          offset="1"
          stop-color="#6366F1"
        />
      </lineargradient>
    </defs>
  </svg>
</div>
`;
